File: Tyche.Android/session/src/main/java/com/felipearpa/tyche/session/authentication/application/LogoutUseCase.kt
package com.felipearpa.tyche.session.authentication.application

import com.felipearpa.tyche.session.AccountStorage
import com.felipearpa.tyche.session.authentication.domain.AuthenticationRepository
import javax.inject.Inject

class LogoutUseCase @Inject constructor(
    private val authenticationRepository: AuthenticationRepository,
    private val accountStorage: AccountStorage
) {
    suspend fun execute(): Result<Unit> {
        val logoutResult = authenticationRepository.logout()
        logoutResult.onSuccess {
            accountStorage.delete()
        }
        return logoutResult
    }
}

File: Tyche.iOS/Session/Sources/Session/Authentication/Application/LogOutUseCase.swift
public class LogOutUseCase {
    private let authenticationRepository: AuthenticationRepository
    private let accountStorage: AccountStorage
    
    init(authenticationRepository: AuthenticationRepository, accountStorage: AccountStorage) {
        self.authenticationRepository = authenticationRepository
        self.accountStorage = accountStorage
    }
    
    func execute() async -> Result<Void, Error> {
        let logoutResult = await authenticationRepository.logout()
        if case .success = logoutResult {
            try! await accountStorage.delete()
        }
        return logoutResult
    }
}
